File: dart/reader.dart
import 'types.dart';

final malRegExp = RegExp(
    r"""[\s,]*(~@|[\[\]{}()'`~^@]|"(?:\\.|[^\\"])*"?|;.*|[^\s\[\]{}('"`,;)]*)""");
final strRegExp = RegExp(r"""^"(?:\\.|[^\\"])*"$""");

class Reader {
  final List<String> tokens;
  int _position = 0;

  Reader(this.tokens);

  String next() {
    var token = peek();
    _position++;
    return token;
  }

  String peek() {
    if (_position >= tokens.length) return null;
    return tokens[_position];
  }
}

class ParseException implements Exception {
  final String message;

  ParseException(this.message);
}

class NoInputException implements Exception {}

MalType read_str(String code) {
  var tokens = tokenizer(code);
  if (tokens.isEmpty) {
    throw NoInputException();
  }
  var reader = Reader(tokens);
  return read_form(reader);
}

List<String> tokenizer(String code) {
  var matches = malRegExp.allMatches(code);
  return matches
      .map((m) => m.group(1))
      .where((token) => token.isNotEmpty && !token.startsWith(';'))
      .toList();
}

MalType read_form(Reader reader) {
  const macros = <String, String>{
    "'": 'quote',
    '`': 'quasiquote',
    '~': 'unquote',
    '~@': 'splice-unquote',
    '@': 'deref',
    '^': 'with-meta',
  };
  const sequenceStarters = <String, String>{'(': ')', '[': ']', '{': '}'};
  var token = reader.peek();
  if (sequenceStarters.containsKey(token)) {
    var elements = read_sequence(reader, token, sequenceStarters[token]);
    if (token == '(') {
      return MalList(elements);
    }
    if (token == '[') {
      return MalVector(elements);
    }

    if (token == '{') {
      return MalHashMap.fromSequence(elements);
    }

    throw StateError("Impossible!");
  } else if (macros.containsKey(token)) {
    var macro = MalSymbol(macros[token]);
    reader.next();
    var form = read_form(reader);
    if (token == '^') {
      var meta = read_form(reader);
      return MalList([macro, meta, form]);
    } else {
      return MalList([macro, form]);
    }
  } else {
    return read_atom(reader);
  }
}

List<MalType> read_sequence(Reader reader, String open, String close) {
  // Consume opening token
  var actualOpen = reader.next();
  assert(actualOpen == open);

  var elements = <MalType>[];
  for (var token = reader.peek();; token = reader.peek()) {
    if (token == null) {
      throw ParseException("expected '$close', got EOF");
    }
    if (token == close) break;
    elements.add(read_form(reader));
  }

  var actualClose = reader.next();
  assert(actualClose == close);

  return elements;
}

MalType read_atom(Reader reader) {
  var token = reader.next();

  var intAtom = int.tryParse(token);
  if (intAtom != null) {
    return MalInt(intAtom);
  }

  if (strRegExp.matchAsPrefix(token) != null) {
    var sanitizedToken = token
        // remove surrounding quotes
        .substring(1, token.length - 1)
        .replaceAllMapped(
            RegExp("\\\\(.)"), (Match m) => m[1] == 'n' ? '\n' : m[1]);
    return MalString(sanitizedToken);
  }

  if (token[0] == '"') {
    throw ParseException("expected '\"', got EOF");
  }

  if (token[0] == ':') {
    return MalKeyword(token.substring(1));
  }

  if (token == 'nil') {
    return MalNil();
  }

  if (token == 'true') {
    return MalBool(true);
  }

  if (token == 'false') {
    return MalBool(false);
  }

  return MalSymbol(token);
}

File: ts/reader.ts
import { MalType, MalList, MalString, MalNumber, MalBoolean, MalNil, MalKeyword, MalSymbol, MalVector, MalHashMap } from "./types";

class Reader {
    position = 0;

    constructor(private tokens: string[]) { }

    next(): string {
        const ret = this.peek();
        this.position += 1;
        return ret;
    }

    peek(): string {
        return this.tokens[this.position];
    }
}

export function readStr(input: string): MalType {
    const tokens = tokenizer(input);
    const reader = new Reader(tokens);
    return readForm(reader);
}

function tokenizer(input: string): string[] {
    const regexp = /[\s,]*(~@|[\[\]{}()'`~^@]|"(?:\\.|[^\\"])*"?|;.*|[^\s\[\]{}('"`,;)]*)/g;
    const tokens: string[] = [];
    while (true) {
        const matches = regexp.exec(input);
        if (!matches) {
            break;
        }
        const match = matches[1];
        if (match === "") {
            break;
        }
        if (match[0] !== ";") {
            tokens.push(match);
        }
    }

    return tokens;
}

function readForm(reader: Reader): MalType {
    const token = reader.peek();
    switch (token) {
        case "(":
            return readList(reader);
        case "[":
            return readVector(reader);
        case "{":
            return readHashMap(reader);
        case "'":
            return readSymbol("quote");
        case "`":
            return readSymbol("quasiquote");
        case "~":
            return readSymbol("unquote");
        case "~@":
            return readSymbol("splice-unquote");
        case "@":
            return readSymbol("deref");
        case "^":
            {
                reader.next();
                const sym = MalSymbol.get("with-meta");
                const target = readForm(reader);
                return new MalList([sym, readForm(reader), target]);
            }
        default:
            return readAtom(reader);
    }

    function readSymbol(name: string) {
        reader.next();
        const sym = MalSymbol.get(name);
        const target = readForm(reader);
        return new MalList([sym, target]);
    }
}

function readList(reader: Reader): MalType {
    return readParen(reader, MalList, "(", ")");
}

function readVector(reader: Reader): MalType {
    return readParen(reader, MalVector, "[", "]");
}

function readHashMap(reader: Reader): MalType {
    return readParen(reader, MalHashMap, "{", "}");
}

function readParen(reader: Reader, ctor: { new (list: MalType[]): MalType; }, open: string, close: string): MalType {
    const token = reader.next(); // drop open paren
    if (token !== open) {
        throw new Error(`unexpected token ${token}, expected ${open}`);
    }
    const list: MalType[] = [];
    while (true) {
        const next = reader.peek();
        if (next === close) {
            break;
        } else if (!next) {
            throw new Error("unexpected EOF");
        }
        list.push(readForm(reader));
    }
    reader.next(); // drop close paren

    return new ctor(list);
}

function readAtom(reader: Reader): MalType {
    const token = reader.next();
    if (token.match(/^-?[0-9]+$/)) {
        const v = parseInt(token, 10);
        return new MalNumber(v);
    }
    if (token.match(/^-?[0-9]\.[0-9]+$/)) {
        const v = parseFloat(token);
        return new MalNumber(v);
    }
    if (token.match(/^"(?:\\.|[^\\"])*"$/)) {
        const v = token.slice(1, token.length - 1)
            .replace(/\\(.)/g, (_, c: string) => c == 'n' ? '\n' : c)
        return new MalString(v);
    }
    if (token[0] === '"') {
        throw new Error("expected '\"', got EOF");
    }
    if (token[0] === ":") {
        return MalKeyword.get(token.substr(1));
    }
    switch (token) {
        case "nil":
            return MalNil.instance;
        case "true":
            return new MalBoolean(true);
        case "false":
            return new MalBoolean(false);
    }

    return MalSymbol.get(token);
}
